File: linked_lists/palindrome/sol2.py
import sys

sys.path.append("../")

from sll import sll

def ispal(slinked_list):
  if slinked_list.size < 3:
    return True
 
  stack_store = []
  
  slow = slinked_list.head
  fast = slow.next
  stack_store.append(slow.data)
  
  while fast and fast.next:
    slow = slow.next
    stack_store.append(slow.data)
    fast = fast.next.next
    
  if not fast:
    stack_store.pop()
    slow = slow.next
  
  slow = slow.next
  
  while slow:
    data = stack_store.pop()
    if data != slow.data:
      return False
    slow = slow.next
  return True
    
if __name__ == "__main__":
  slinked_list1 = sll.sll()
  key = 0
  value = [7]
  for it in value:
    slinked_list1.addLast(key, it)
    key += 1
  
  print ispal(slinked_list1)
 

File: linked_lists/palindrome/sol2.cpp
#include "sll.h"

#include <iostream>
#include <vector>


node* reverseLinkedList(node* nd) {
  if ( !nd ) return nd;
  node* tail = nd;
  node* head = nd;
  
  while ( tail->next() ) {
    node* curr = tail->next();
    tail->setNext(curr->next());
    curr->setNext(head);
    head = curr;
  }
  return head;
}


bool isPalindrome(node* nd) {
  node* slow = nd;
  node* fast = nd;
  
  while ( fast && fast->next() ) {
     slow = slow->next();
     fast = fast->next()->next();
  }
  
  if ( fast ) {
    slow = slow->next();
  }
  
  slow = reverseLinkedList(slow);
  
  fast = nd;
  while ( fast && slow ) {
    if ( fast->data() != slow->data() ) {
      return false;
    }
    fast = fast->next();
    slow = slow->next();
  }
  
  return true;
}
  

int main() {
  sll slinked_list;
  std::vector<int> value{1, 2, 3, 2, 1};
  int key = 0;
  for (auto& it : value) {
    if (key == 0) {
      slinked_list.insertHead(it, key);
    } else {
      slinked_list.insertLast(it, key);
    }
    ++key;
  }
  
  node* nd = slinked_list.head();
  while (nd) {
    std::cout << nd->data() << " ";
    nd = nd->next();
  }
  std::cout << std::endl;
   

  nd = slinked_list.head();
  if ( isPalindrome(nd) ) {
     std::cout << "Palindrome";
  } else {
     std::cout << "is not";
  }
  
  std::cout << std::endl;
}
